### PR TITLE
Allow to ignore duplicate entires when adding files to SQL project.

### PR DESCRIPTION
**Pull request by nahk-ivanov**
This change introduces new boolean flag for `addToProject`, `addScriptItem` and `addFolderItem` APIs that allows to skip files/folders if they are already included in the project file. Original behavior was to throw an error if duplicate entry is encountered, so I kept that as a default (new flag is not set). This new flag allows to reuse the de-dupe code that we otherwise have to add to each consumer of the API.

I started by fixing the original behavior, as it was not always working as expected. In our extension that relies on `addToProject` API we've hit an issue where duplicate items were added and no errors were thrown. There was a test for this scenario, but unfortunately the test had few bugs in it as well, so I addressed that first. First issue with the test code was missing `await` on the `testUtils.shouldThrowSpecificError` calls, so test was not actually testing anything. After adding missing keywords, I hit an issue where exception was not thrown, but this turned out to be different issue, compared to what we were hitting. In the test code, it was using the very first folder from the generate list of test entries. This folder was actually the root of the project (where sqlproj file is located), and `addToProject` API had a special case to ignore the root. This means neither first call nor second call was actually adding anything to the project and no errors were produced. I fixed this problem by using next available folder in the generated files list.

After addressing the test code I could not reproduce the issue that we were seeing with duplicate entries being added, everything was working as expected and errors were thrown. I started adding more tests that better resemble our production scenario - add files in subfolders, add files to existing project, rather than a new one. Finally I was able to reproduce the problem in tests when adding a file in a subfolder to an existing project. After investigation this turned out to be an issue with mismatch in how `relativePath` is maintained within the `FileProjectEntry`. When loading an existing project, `relativePath` is populated based on the value of the `Include` attribute of the `Build` item. This attribute is normalized to Windows-style path, using `\`, so for nested file you will have `folder\file.sql`. When adding new item to the project, one could pass either Windows or Unix-style path (`folder/file.sql`), so the path comparison between loaded Windows-style path and newly added Unix-style path was failing, resulting in them being treated as different items. In fact, `addToProject` API that we were using was relying on `Uri` helpers to extract relative path, thus the path was forced to be Unix-style and path was never the same as the loaded one. After this discovery I added a dedicated test to validate the round-trip of the `relativePath` for serialized and desirialized project file.

In order to address this problem, I updated the factory method `createFileProjectEntry` to always run `utils.convertSlashesForSqlProj` on the relative path to ensure we have Windows-style path stored in there. I also optimized the helper code slightly to not do split/join, if there are no split points in the input string, which should eliminate unnecessary array instantiation. It is worth mentioning that I had to normalize the input relative paths in the `addScriptItem` and `addFolderItem` APIs, because there is no guarantee that they will be Windows-style when we try to compare them to `relativePath` of the existing project items.

Finally I was able to add a simply flag and update the condition to return existing record, if duplicates were allowed. I also updated typings file for the extension and added tests to cover this scenario.

### CODE REVIEW

**Comment by nahk-ivanov** [comment]
I should also mention that treating `buildElements[b].getAttribute(constants.Include)` as a relative path to the item is not correct to begin with: `Include` attribute is an expression, not a path, it can be `file1.sql;file2.sql` or `**\*.sql` with `Exclude="**\obj\*.sql"`. You need to run msbuild to evaluate the item groups and I did not address that issue with this change.

File: extensions/sql-database-projects/src/models/project.ts
Comment (nahk-ivanov) [inline extensions/sql-database-projects/src/models/project.ts:319]: I changed comparison to use upper-case to make sure non-Latin characters are compared properly.

Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/models/project.ts:319]: Conversion should also be done outside of the comparison delegate, as it is a constant value, but the code is already plagued with bad coding practices, so I went for a least destructive change to simplify diff.

Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/models/project.ts:319]: I also don't know if there was a reason to use property getter (`this.files`), instead of a field directly (`this._files`). I changed it to use the field, as done in other places in this file.

File: extensions/sql-database-projects/src/models/project.ts
Comment (nahk-ivanov) [inline extensions/sql-database-projects/src/models/project.ts:314]: I was debating whether to use default value or `boolean | undefined` type. It appears that `.. | undefined` is the preferred way in this codebase, so I followed that practice.

File: extensions/sql-database-projects/src/models/project.ts
Comment (nahk-ivanov) [inline extensions/sql-database-projects/src/models/project.ts:578]: Technically I could fix our problem without adding `utils.convertSlashesForSqlProj` call here, as I already updated `addScriptItem` and `addFolderItem` to normalize the path for comparison, but I felt like this would make it a little bit less error-prone in case there are other places in the future calling this.

I also considered adding this normalization in the `FileProjectEntry` constructor directly, to solve the problem once and for all, but after a closer look I assumed that `FileProjectEntry` was designed as DTO, so I didn't want to break that pattern by introducing logic in there.

File: extensions/sql-database-projects/src/test/project.test.ts
Comment (nahk-ivanov) [inline extensions/sql-database-projects/src/test/project.test.ts:579]: I'm not fond of magic numbers, but I didn't want to butcher the code too much for the sake of diff. The array is as following:
```javascript
[
  '', // [0]
  'script1.sql', // [1]
  'folder1', // [2]
  'folder1\script1.sql', // [3]
  ...
]
```

File: extensions/sql-database-projects/src/models/project.ts
Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/models/project.ts:578]: There is actually another issue here as well which I did not address - trailing `\`. i.e. `folder1` vs. `folder1\`. It will not surface if you only use `addToProject`, but if you use `addFolderItem` or create an entry using this factory method directly, then you may hit it. We can update `convertSlashesForSqlProj` helper to trim `\` maybe? I think the purpose of the method was to normalize the values for `sqlproj` file, right?

**Comment by Benjin** [comment]
I'm missing something fundamental about this delta... what are the situations where we would ever _want_ duplicates added to the project's list of files (and not ignore them)?
I think a better model would be for the caller of `addFile()` to check to make sure whatever they're adding isn't a duplicate via a file list property first (or dedupe their list of to-be-added files) before making the call.  For a situation where this selection isn't controlled by us (e.g. a user adding an existing file, or typing in a duplicate file name), it'll detect the duplicate, throw an error, and that error would then be caught and handled appropriately by the caller (descriptive message/toast).

**Comment by nahk-ivanov** [comment]
@Benjin 

> what are the situations where we would ever want duplicates added to the project's list of files (and not ignore them)?

I don't know, probably none (although Visual Studio allows it and displays them properly as two items). This change is not enabling that scenario, but rather prevents it - before this change it was possible to have duplicates added to the project when calling `addToProject` (bug).

> I think a better model would be for the caller of addFile() to check to make sure whatever they're adding isn't a duplicate via a file list property first (or dedupe their list of to-be-added files) before making the call.

This is exactly what we are trying to avoid, as description for the change clearly states:

> This new flag allows to reuse the de-dupe code that we otherwise have to add to each consumer of the API.

Code duplication, in general, is not considered a "better model" by any stretch. Few options we have here is:
1) Have separate package with helper methods that would implement such common functionality as de-dupe
2) Extract de-dupe logic to a separate helper class that can be exposed by this extension through factory method or as a static instance

After looking at the code, I noticed that there was already a validation for duplicate items, which would run for each file-add operation anyway, so I decided it will be more efficient in terms of performance to rely on the existing validation and just allow the caller to control the behavior - whether you want an error or silently continue, i.e. `TryParse` vs `Parse`, PowerShell's `-ErrorAction`, etc.

Typical reason why this is preferred is a typical TOCTOU race condition - unless project API allows an exclusive lock of the project, there will be no way for the caller to actually safely check whether item is duplicate or not within the single ADS session itself, as another caller can add a file after we check that file is not there, therefore *better model* is usually considered to rely on exceptions/return codes rather than "check now and act later". The implementation in this change allows that. We could probably run `addScriptItem`/`addFolderItem` directly for each and every item and catch the error, but ultimately it would be more efficient to have the same "silently ignore" flag instead of constantly throwing and catching.

Whether you want to prevent the TOCTOU between your extension and file system - is up to you. I'd assume you do want that based on what you said - there are probably no cases where we want to have duplicates. I believe there is another workitem already somewhere to throw if project has changed after it was read by the extension, so it should cover it, me thinks. I did not touch *extension to file system* communication, as we act as a caller for this API and my main focus was to ensure that between our extensions and your public API surface we are somewhat safe.

File: extensions/sql-database-projects/src/sqldbproj.d.ts
Comment (Benjin) [inline extensions/sql-database-projects/src/sqldbproj.d.ts:40]: whether duplicate entries should be ignored or throw an error.

(and elsewhere)

File: extensions/sql-database-projects/src/test/project.test.ts
Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/test/project.test.ts:579]: You don't want to hardcode either. What we want here is to actually find first/second/third/etc. folder/file item in the array using `fs.stat` or something. That will be a slightly bigger code change, which I didn't want to include in this changeset, as it already takes 3+ days to get this in and we would like to move a little bit faster than that. I will keep it this way for now just for the sake of making progress and you can add a workitem to improve this within your team.

File: extensions/sql-database-projects/src/sqldbproj.d.ts
Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/sqldbproj.d.ts:42]: I did consider using the inverted name (i.e. `throwIfDuplicate`), but decided against it, because to preserve existing default behavior it would have to be set to `true`. Given that this code base prefers `T | undefined` instead of explicit defaults and `true` is rarely a default for boolean type, I decided to use the non-inverted name that aligns with the new feature that is being added, rather than original default behavior. This also exposes lack of default values for interfaces in TypeScript with all consequences, i.e. one cannot do `throwIfDuplicate: boolean = true` for an interface method, so interface will be `throwIfDuplicate?: boolean`, but class will be `throwIfDuplicate: boolean = true`.
After all, I don't see how the inverted name solves the suggested "ambiguity", as it still does not indicate what is the alternative (to throwing, in that case) - skip or add. I can suggest `doNotThrowOnDuplicate`, which will address the above and presumably eliminate the confusion around failing behavior, which, being the original default, I have never thought was a confusion at first place.

I'm not really sure if this deserves an enumeration, given that there are just two values at the moment. This can be refactored later if additional duplicate handling strategies are possible (which they are). Keep in mind that enumeration also has the problem with obscure default, where a boolean flag with `false` default is less so.

Comment (Benjin) [reply extensions/sql-database-projects/src/sqldbproj.d.ts:42]: I agree that it an enum would be overkill here, but it would also leave the least open to different interpretations.  `doNotThrowOnDuplicate` is fine as well.

File: extensions/sql-database-projects/src/models/project.ts
Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/models/project.ts:319]: I'm not really sure how .NET is in play here, as this is a culture-related issue, where in some cultures there might be two upper-case characters (accented and not), but only one low-case character, thus any implementation of to-lower will lose fidelity. I don't think `string.normalize()` should be used here either. If we want a CI AS comparison, then `Intl.Collator` with appropriate configuration should be used.

In reality though, what we want is not a CI AS comparison, but to determine whether this file on disk is already added to the project and this is a little bit more complicated than just running a string comparison. First of all, it is based on the file system where the file is stored - some allow same name with different case and some do not. Then there is an extra normalization necessary for folders to make sure correct path separator is used and trailing separators are present/trimmed for folders. Ultimately you probably want to find a file/folder on disk using `fs.stat` or something and extract canonical file name from it for comparison. This further hints the necessity of this comparison logic to be present here, rather than duplicated within each and every consumer.

Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/models/project.ts]: > Depends on the codebase's priorities. For performance, yes. For concise, more-readable code, something like this is better.

:) This is probably the most bizarre statement I've ever seen. I'd pay to see a conversation with a customer where we explain to them that our tools are running slow, because instead our code looks "pretty" and that what matters the most.
Not trying to diminish the value of readable and self-explanatory code, but I don't think there is ever a choice for "codebase's priorities" between user experience (performance, in this case) and code readability. Usually you would decide on the target user experience and try to achieve it with the most "concise, more-readable code" possible, but not the opposite. The point can be made that the performance is "good enough", but I think this is one of the cases where improving it doesn't require virtually any effort. Otherwise (i.e. "we must have it written this way, otherwise I don't understand what is going on"), it probably deserve a change in `tsc` to improve the DFA and optimize this during transpilation.

File: extensions/sql-database-projects/src/sqldbproj.d.ts
Comment (nahk-ivanov) [reply extensions/sql-database-projects/src/sqldbproj.d.ts]: Renamed the argument to `doNotThrowOnDuplicate`. Hopefully nobody complains that it's now unclear what is the behavior when we do not throw. 🤦

**Review by Benjin — APPROVED** [review]

**Comment by Benjin** [comment]
Everyone indicated as a "code owner" in the reviewer box (as indicated by the keyed shield icon), so once @kisantia signs off, you're good to merge.

**Comment by nahk-ivanov** [comment]
> once @kisantia signs off, you're good to merge

I don't think this is correct, Kim did sign off, but I'm still unable to merge:

image

I assume it has to be one of you? What's the process? Last person with the "keyed shield icon" to sign off merges the PR? Or should I request write permissions? I don't see any __Merge__ buttons anywhere.

**Comment by Charles-Gagnon** [comment]
@nahk-ivanov I've added you to the contrib team so you should be able to merge in now.

**Comment by nahk-ivanov** [comment]
@Charles-Gagnon thanks! Does this imply that I can now create branches within this repo, instead of maintaining my own fork or do you guys still prefer forks?

**Comment by Charles-Gagnon** [comment]
Yeah you can create branches now - and we definitely prefer branches within the repo since that makes it a lot easier for us to pull down a branch to look at something if we need to take a closer look at anything you're working on.

**Pull request merged**
